doc/charm++/manual.rst

```diff
@@ -4811,55 +4854,187 @@ registration memory handles and will not incur any registration costs.
+   // or arrayProxy.foo(500000, CkSendBuffer(arrPtr, CK_BUFFER_UNREG)); for UNREG reg-mode
+   // or arrayProxy.foo(500000, CkSendBuffer(arrPtr, CK_BUFFER_PREREG)); for PREREG reg-mode
+
+   /* Specifying de-registration modes */
+
+   // for point to point send
+   arrayProxy[0].foo(500000, CkSendBuffer(arrPtr, CK_BUFFER_REG, CK_BUFFER_DEREG));
+
+   // or arrayProxy[0].foo(500000, CkSendBuffer(arrPtr, CK_BUFFER_REG)); for DEREG dereg-mode because of DEREG being the default
+   // or arrayProxy[0].foo(500000, CkSendBuffer(arrPtr, CK_BUFFER_REG, CK_BUFFER_NODEREG)); for NODEREG dereg-mode
+
+   // for bcast send
+   arrayProxy.foo(500000, CkSendBuffer(arrPtr, CK_BUFFER_REG, CK_BUFFER_DEREG));
+
+   // or arrayProxy.foo(500000, CkSendBuffer(arrPtr, CK_BUFFER_REG)); for DEREG dereg-mode because of DEREG being the default
+   // or arrayProxy.foo(500000, CkSendBuffer(arrPtr, CK_BUFFER_REG, CK_BUFFER_NODEREG)); for NODEREG dereg-mode
+
+
+In the case of point to point communication, the callback will be invoked on completion
+of the RDMA operation associated with the corresponding array.
+
+For zero copy broadcast operations, the callback will be invoked when all the broadcast
+recipient processes have successfully received the broadcasted array. Note that this does
+not necessarily mean that the recipient objects have received the data i.e. it is possible that the
+process (or the logical node) has received the data, but the individual entry methods
+are still waiting in the scheduler (and are scheduled to run) and hence have not received the
+data.
+
+
+Inside the callback, it is safe to overwrite the buffer sent via the Zero Copy Entry Method Send API.
+
+.. code-block:: c++
+
+   //called when RDMA operation is completed
+   void zerocopySent() {
+     // update the buffer to the next pointer
+     updateBuffer();
+   }
 
-The callback will be invoked on completion of the RDMA operation
-associated with the corresponding array. Inside the callback, it is safe
-to overwrite the buffer sent via the zero copy entry method send API and
-this buffer can be accessed by dereferencing the CkDataMsg received in
-the callback.
+The callback methods can also take a pointer to a ``CkDataMsg`` message.
+This message can be used to access the original buffer information
+object i.e. the ``CkNcpyBuffer`` object constructed by CkSendBuffer and
+used for the zero copy transfer. The buffer information object available in
+the callback allows access to all its information including the buffer pointer.
 
 .. code-block:: c++
 
    //called when RDMA operation is completed
-   void zerocopySent(CkDataMsg *m)
+   void zerocopySent(CkDataMsg *msg)
    {
-     //get access to the pointer and free the allocated buffer
-     void *ptr = *((void **)(m->data));
-     free(ptr);
-     delete m;
+     // Cast msg->data to a CkNcpyBuffer to get the source buffer information object
+     CkNcpyBuffer *source = (CkNcpyBuffer *)(m->data));
```

Comment:
should be msg->data I think

Reply:
Good catch. Fixed in this PR (https://github.com/UIUC-PPL/charm/pull/2157)